### PR TITLE
Update CRDs as per GEP-3155

### PR DESCRIPTION
**Pull request by mkosieradzki**
/kind gep
**What this PR does / why we need it**:
Updates CRDs as per GEP-3155

**Does this PR introduce a user-facing change?**:
```release-note
Updated CRDs as per GEP-3155
```

### CODE REVIEW

File: apis/v1alpha3/backendtlspolicy_types.go [hunk 1]
```diff
@@ -126,11 +141,50 @@ type BackendTLSPolicyValidation struct {
 	// backends:
 	//
 	// 1. Hostname MUST be used as the SNI to connect to the backend (RFC 6066).
-	// 2. Hostname MUST be used for authentication and MUST match the certificate
-	//    served by the matching backend.
+	// 2. Only if SubjectAltNames is not specified, Hostname MUST be used for
```
Comment (robscott) [inline apis/v1alpha3/backendtlspolicy_types.go:144]: Tiny nit:
```suggestion
	// 2. If SubjectAltNames is not specified, Hostname MUST be used for
```

File: apis/v1alpha3/backendtlspolicy_types.go [hunk 2]
```diff
@@ -126,11 +141,50 @@ type BackendTLSPolicyValidation struct {
...
+	// Support: Core
+	Type SubjectAltNameType `json:"type"`
+
+	// Hostname contains Subject Alternative Name specified in DNS name format.
+	// Required when Type is set to Hostname, ignored otherwise.
+	//
+	// Support: Core
+	//
+	// +optional
+	Hostname v1.PreciseHostname `json:"hostname,omitempty"`
```
Comment (robscott) [inline apis/v1alpha3/backendtlspolicy_types.go:179]: Not an expert here, but it seems like a wildcard hostname is at least theoretically possible (https://stackoverflow.com/a/21496451). That would mean using the more generic type that also allows wildcards:

https://github.com/kubernetes-sigs/gateway-api/blob/f7d4a8715ad831d84fd3142214ca2f54647ec9e8/apis/v1/shared_types.go#L523

File: apis/v1alpha3/backendtlspolicy_types.go [hunk 3]
```diff
@@ -126,11 +141,50 @@ type BackendTLSPolicyValidation struct {
...
+
+	// URI contains Subject Alternative Name specified in URI format.
+	// Required when Type is set to URI, ignored otherwise.
+	//
+	// Support: Core
+	//
+	// +optional
+	URI string `json:"uri,omitempty"`
```
Comment (robscott) [inline apis/v1alpha3/backendtlspolicy_types.go:187]: Seems like we need some kind of regex validation here, similar in concept to what we've added for the hostname types.

File: apis/v1alpha3/backendtlspolicy_types.go [hunk 4]
```diff
@@ -126,11 +141,50 @@ type BackendTLSPolicyValidation struct {
...
+	// +optional
+	// +kubebuilder:validation:MaxItems=5
+	SubjectAltNames []SubjectAltName `json:"subjectAltNames,omitempty"`
+}
+
+// SubjectAltName represents Subject Alternative Name.
+// +kubebuilder:validation:XValidation:message="must contain Hostname, if Type is set to Hostname",rule="!(self.Type == "Hostname" && has(self.Hostname) && self.Hostname != \"\")"
+// +kubebuilder:validation:XValidation:message="must not contain Hostname, if Type is not set to Hostname",rule="self.Type != "Hostname" && has(self.Hostname) && self.Hostname != \"\""
+// +kubebuilder:validation:XValidation:message="must contain URI, if Type is set to URI",rule="!(self.Type == "URI" && has(self.URI) && self.HostName != \"\")"
+// +kubebuilder:validation:XValidation:message="must not contain URI, if Type is not set to URI",rule="self.Type != "URI" && has(self.URI) && self.URI != \"\""
```
Comment (robscott) [inline apis/v1alpha3/backendtlspolicy_types.go:166]: Recommend adding some basic CEL test coverage, similar to https://github.com/kubernetes-sigs/gateway-api/pull/3283

**Review by robscott — COMMENTED** [review]
Thanks @mkosieradzki!

**Comment by LiorLieberman** [comment]
this moved to #3304

**Pull request closed without merging**